test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,357 @@
+				}
+
+				// Ensure deleted objects are gone
+				for _, obj := range objectsDeleted {
+					err := input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(obj), obj)
+					g.Expect(err).To(HaveOccurred())
+					g.Expect(apierrors.IsNotFound(err)).To(BeTrue())
+				}
+
+				// Ensure other objects are not in deletion.
+				for _, obj := range objectsNotInDeletion {
+					g.Expect(input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(obj), obj)).ToNot(HaveOccurred())
+					g.Expect(obj.GetDeletionTimestamp().IsZero()).To(BeTrue())
+				}
+			}).WithTimeout(5 * time.Minute).WithPolling(10 * time.Second).Should(Succeed())
+
+			// Remove the test's finalizer from all objects which had been expected to be in deletion to unblock the next phase.
+			removeFinalizer(ctx, input.BootstrapClusterProxy.GetClient(), finalizer, objectsInDeletion...)
+		}
+
+		// The last phase should unblock the cluster to actually have been removed.
+		log.Logf("Waiting for the Cluster %s to be deleted", klog.KObj(clusterResources.Cluster))
+		framework.WaitForClusterDeleted(ctx, framework.WaitForClusterDeletedInput{
+			Client:         input.BootstrapClusterProxy.GetClient(),
+			Cluster:        clusterResources.Cluster,
+			ArtifactFolder: input.ArtifactFolder,
+		}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		// Dump all the resources in the spec namespace and the workload cluster.
+		framework.DumpAllResourcesAndLogs(ctx, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, clusterResources.Cluster)
+
+		if !input.SkipCleanup {
+			// Remove finalizers we added to block normal deletion.
+			removeFinalizer(ctx, input.BootstrapClusterProxy.GetClient(), finalizer, objectsWithFinalizer...)
+
+			By(fmt.Sprintf("Deleting cluster %s", klog.KObj(clusterResources.Cluster)))
+			// While https://github.com/kubernetes-sigs/cluster-api/issues/2955 is addressed in future iterations, there is a chance
+			// that cluster variable is not set even if the cluster exists, so we are calling DeleteAllClustersAndWait
+			// instead of DeleteClusterAndWait
+			framework.DeleteAllClustersAndWait(ctx, framework.DeleteAllClustersAndWaitInput{
+				Client:         input.BootstrapClusterProxy.GetClient(),
+				Namespace:      namespace.Name,
+				ArtifactFolder: input.ArtifactFolder,
+			}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+
+			By(fmt.Sprintf("Deleting namespace used for hosting the %q test spec", specName))
+			framework.DeleteNamespace(ctx, framework.DeleteNamespaceInput{
+				Deleter: input.BootstrapClusterProxy.GetClient(),
+				Name:    namespace.Name,
+			})
+		}
+		cancelWatches()
+	})
+}
+
+func addFinalizer(ctx context.Context, c client.Client, finalizer string, objs ...client.Object) {
+	for _, obj := range objs {
+		Expect(c.Get(ctx, client.ObjectKeyFromObject(obj), obj)).To(Succeed())
```

Comment:
nit: let's add error message for when the assertion fails
(same for other Expect)